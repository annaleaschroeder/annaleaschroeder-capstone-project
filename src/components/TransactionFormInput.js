import React from 'react'
import styled from 'styled-components/macro'
import { useForm } from 'react-hook-form'
import PropTypes from 'prop-types'

TransactionFormInput.propTypes = {
  onSave: PropTypes.func.isRequired,
}

export default function TransactionFormInput({ onSave, value, setValue }) {
  const { register, handleSubmit, errors } = useForm()

  const onSubmit = (event) => {
    //required for testing, otherwise line 15 and 16 are redundand and could be deleted
    if (event && event.target && typeof event.target.reset === 'function')
      event.target.reset()

    onSave()
  }

  const onChange = (event) => {
    setValue(event.target.value)
  }

  return (
    <FormStyled onSubmit={handleSubmit(onSubmit)}>
      <InputContainer>
        <input
          id="transactionInput"
          name="transactionInput"
          placeholder="Enter new Transaction"
          onChange={onChange}
          value={value}
          ref={register({
            required: true,
            min: 1,
            pattern: /^[0-9]+,[0-9]{2}$/,
          })}
        />
        {errors.transactionInput &&
          errors.transactionInput.type === 'required' && (
            <span>Transaction required. Example: 20,95.</span>
          )}
        {errors.transactionInput &&
          errors.transactionInput.type === 'pattern' && (
            <span>Please use the following format: 12,05</span>
          )}
      </InputContainer>
      <LableStyled htmlFor="transactionInput">Euro</LableStyled>
      {/* <CancelButton type="reset">Cancel</CancelButton> */}
      <AddTransactionButton type="submit">Add Transaction</AddTransactionButton>
    </FormStyled>
  )
}

const FormStyled = styled.form`
  display: grid;
  justify-items: center;
  text-align: center;
  gap: 20px;
  padding: 20px;
`
const InputContainer = styled.div`
  display: flex;
`

const AddTransactionButton = styled.button`
  display: block;
  width: 200px;
  height: auto;
  padding: 10px;
  background-color: lightgrey;
  justify-content: grid-column-end;
`

const LableStyled = styled.label`
  font-weight: bold;
`
// const CancelButton = styled.button`
//   display: block;
//   width: 100px;
//   height: auto;
//   padding: 10px;
//   background-color: lightgrey;
//   justify-content: grid-column-end;
// `
