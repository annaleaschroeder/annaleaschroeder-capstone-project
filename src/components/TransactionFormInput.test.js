import React from 'react'
import TransactionFormInput from './TransactionFormInput'
import renderer from 'react-test-renderer'
import { render, fireEvent, waitFor } from '@testing-library/react'

import 'jest-styled-components'

window.MutationObserver = require('mutation-observer')

describe('TransactionInputForm', () => {
  it('creates a new transaction', async () => {
    const onSave = jest.fn()
    const { getByText, getByLabelText } = render(
      <TransactionFormInput onSave={onSave} />
    )

    fireEvent.change(getByLabelText('Euro'), {
      target: { value: '15,95' },
    })

    fireEvent.submit(getByText('Add Transaction'))
    await waitFor(() => expect(onSave).toHaveBeenCalled())
  })

  it('Displays a form to create new transactions', async () => {
    const onSave = jest.fn()
    const { getByPlaceholderText } = render(
      <TransactionFormInput onSave={onSave} />
    )

    const transactionInput = getByPlaceholderText('Enter new Transaction')
    expect(transactionInput.value).toBe('')
    fireEvent.input(transactionInput, { target: { value: '15,95' } })
    expect(transactionInput.value).toBe('15,95')
  })

  it('renders correctly', () => {
    const onSave = jest.fn()
    const tree = renderer.create(<TransactionFormInput onSave={onSave} />)
    expect(tree).toMatchSnapshot()
  })
})
