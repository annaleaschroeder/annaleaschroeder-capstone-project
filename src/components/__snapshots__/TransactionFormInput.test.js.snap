// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionInputForm renders correctly 1`] = `
.c0 {
  display: grid;
  justify-items: center;
  text-align: center;
  gap: 20px;
  padding: 20px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c3 {
  display: block;
  width: 200px;
  height: auto;
  padding: 10px;
  background-color: lightgrey;
  -webkit-box-pack: grid-column-end;
  -webkit-justify-content: grid-column-end;
  -ms-flex-pack: grid-column-end;
  justify-content: grid-column-end;
}

.c2 {
  font-weight: bold;
}

<form
  className="c0"
  onSubmit={[Function]}
>
  <div
    className="c1"
  >
    <input
      id="transactionInput"
      name="transactionInput"
      onChange={[Function]}
      placeholder="Enter new Transaction"
    />
  </div>
  <label
    className="c2"
    htmlFor="transactionInput"
  >
    Euro
  </label>
  <button
    className="c3"
    type="submit"
  >
    Add Transaction
  </button>
</form>
`;
